File: 2020_2학기_알고리즘/3주차/3.py
import sys
import copy
sys.setrecursionlimit(10**6)

n,m = map(int, input().split(' '))
x = int(input())
idx = 0
arr = [[0 for _ in range(n)] for _ in range(x + 2)]

def ret():
	global idx,n
	i = 0
	while i < (n + 1) // 2:
		print(arr[idx][i],end='')
		i = i + 1
	if n % 2 == 0 :
		i = i - 1
	else:
		i =i - 2
	while i >= 0:
		print(arr[idx][i],end='')
		i -= 1
	sys.exit()

def func(j):
	global idx,n,m
	if j == (n + 1) // 2 :
		arr[idx + 1] = copy.deepcopy(arr[idx])
		if idx == x - 1:
			ret()
		idx = idx + 1
		return
	for i in range(m):
		arr[idx][j] = chr(97 + i)
		func(j + 1)

func(0)

File: 2020_2학기_알고리즘/3주차/3.cpp
#include <iostream>
#include <cstring>
using namespace std;

char arr[100001][200] = {0, };
int n, m, x;
int idx = 0;

void print(){
	int i = 0;
	for (i = 0; i < (n + 1) / 2; i++){
		cout << arr[idx][i];
	}
	i = n % 2 == 0 ? i - 1: i - 2;

	while (i >= 0){
		cout << arr[idx][i];
		i--;
	}
	cout << endl;
	exit(0);
}

void func(int j){
	if (j == (n + 1) / 2){
		strcpy(arr[idx + 1], arr[idx]);
		if (idx == x - 1)
			print();
		idx++;
		return ;
	}

	for (int i = 0; i < m; i++){
		arr[idx][j] = 'a' + i;
		func(j + 1);
	}
}

int main(){
	cin >> n >> m;
	cin >> x;
	func(0);
}
